### ISSUE TITLE
Test failure with test_data_out_of_range on 32-bit 

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#7127]
The 32-bit CircleCI tests in https://github.com/astropy/astropy/pull/7126 have shown a new issue/test failure with the fast reader in io.ascii

Click 'details' show the test log: <details>
```
_______________________ test_data_out_of_range[False-1] ________________________
[gw2] linux -- Python 3.6.4 /opt/python/cp36-cp36m/bin/python

parallel = False, reader = 1

    @pytest.mark.parametrize('reader', [0, 1, 2])
    # catch Windows environment since we cannot use _read() with custom fast_reader
    @pytest.mark.parametrize("parallel", [False,
        pytest.param(True, marks=pytest.mark.xfail(os.name == 'nt', reason="Multiprocessing is currently unsupported on Windows"))])
    def test_data_out_of_range(parallel, reader):
        """
        Numbers with exponents beyond float64 range (|~4.94e-324 to 1.7977e+308|)
        shall be returned as 0 and +-inf respectively by the C parser, just like
        the Python parser.
        Test fast converter only to nominal accuracy.
        """
        # Python reader and strtod() are expected to return precise results
        rtol = 1.e-30
        if reader > 1:
            rtol = 1.e-15
        # passing fast_reader dict with parametrize does not work!
        if reader > 0:
            fast_reader = {'parallel': parallel, 'use_fast_converter': reader > 1}
        else:
            fast_reader = False
        if parallel:
            if reader < 1:
                pytest.skip("Multiprocessing only available in fast reader")
            elif TRAVIS:
                pytest.xfail("Multiprocessing can sometimes fail on Travis CI")
    
        fields = ['10.1E+199', '3.14e+313', '2048e+306', '0.6E-325', '-2.e345']
        values = np.array([1.01e200, np.inf, np.inf, 0.0, -np.inf])
        t = ascii.read(StringIO(' '.join(fields)), format='no_header', guess=False,
                       fast_reader=fast_reader)
        read_values = np.array([col[0] for col in t.itercols()])
        assert_almost_equal(read_values, values, rtol=rtol, atol=1.e-324)
    
        # test some additional corner cases
        fields = ['.0101E202', '0.000000314E+314', '1777E+305', '-1799E+305', '0.2e-323',
                   '2500e-327', ' 0.0000000000000000000001024E+330']
        values = np.array([1.01e200, 3.14e307, 1.777e308, -np.inf, 0.0, 4.94e-324, 1.024e308])
        t = ascii.read(StringIO(' '.join(fields)), format='no_header', guess=False,
                       fast_reader=fast_reader)
        read_values = np.array([col[0] for col in t.itercols()])
>       assert_almost_equal(read_values, values, rtol=rtol, atol=1.e-324)

astropy/io/ascii/tests/test_c_reader.py:997: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
astropy/io/ascii/tests/common.py:42: in assert_almost_equal
    assert np.allclose(a, b, **kwargs)
/opt/python/cp36-cp36m/lib/python3.6/site-packages/numpy/core/numeric.py:2256: in allclose
    res = all(isclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = array(['1.01e+200', '3.14e+307', '1.777e+308', '-inf', '0.2e-323',
       '5e-324', '1.024e+308'], dtype='<U32')
b = array([1.010e+200, 3.140e+307, 1.777e+308,       -inf, 0.000e+000,
       4.941e-324, 1.024e+308])
rtol = 1e-30, atol = 0.0, equal_nan = False

    def isclose(a, b, rtol=1.e-5, atol=1.e-8, equal_nan=False):
        """
        Returns a boolean array where two arrays are element-wise equal within a
        tolerance.
    
        The tolerance values are positive, typically very small numbers.  The
        relative difference (`rtol` * abs(`b`)) and the absolute difference
        `atol` are added together to compare against the absolute difference
        between `a` and `b`.
    
        Parameters
        ----------
        a, b : array_like
            Input arrays to compare.
        rtol : float
            The relative tolerance parameter (see Notes).
        atol : float
            The absolute tolerance parameter (see Notes).
        equal_nan : bool
            Whether to compare NaN's as equal.  If True, NaN's in `a` will be
            considered equal to NaN's in `b` in the output array.
    
        Returns
        -------
        y : array_like
            Returns a boolean array of where `a` and `b` are equal within the
            given tolerance. If both `a` and `b` are scalars, returns a single
            boolean value.
    
        See Also
        --------
        allclose
    
        Notes
        -----
        .. versionadded:: 1.7.0
    
        For finite values, isclose uses the following equation to test whether
        two floating point values are equivalent.
    
         absolute(`a` - `b`) <= (`atol` + `rtol` * absolute(`b`))
    
        The above equation is not symmetric in `a` and `b`, so that
        `isclose(a, b)` might be different from `isclose(b, a)` in
        some rare cases.
    
        Examples
        --------
        >>> np.isclose([1e10,1e-7], [1.00001e10,1e-8])
        array([True, False])
        >>> np.isclose([1e10,1e-8], [1.00001e10,1e-9])
        array([True, True])
        >>> np.isclose([1e10,1e-8], [1.0001e10,1e-9])
        array([False, True])
        >>> np.isclose([1.0, np.nan], [1.0, np.nan])
        array([True, False])
        >>> np.isclose([1.0, np.nan], [1.0, np.nan], equal_nan=True)
        array([True, True])
        """
        def within_tol(x, y, atol, rtol):
            with errstate(invalid='ignore'):
                return less_equal(abs(x-y), atol + rtol * abs(y))
    
        x = asanyarray(a)
        y = asanyarray(b)
    
        # Make sure y is an inexact type to avoid bad behavior on abs(MIN_INT).
        # This will cause casting of x later. Also, make sure to allow subclasses
        # (e.g., for numpy.ma).
        dt = multiarray.result_type(y, 1.)
        y = array(y, dtype=dt, copy=False, subok=True)
    
>       xfin = isfinite(x)
E       TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

/opt/python/cp36-cp36m/lib/python3.6/site-packages/numpy/core/numeric.py:2332: TypeError
________________________ test_data_out_of_range[True-1] ________________________
[gw0] linux -- Python 3.6.4 /opt/python/cp36-cp36m/bin/python

parallel = True, reader = 1

    @pytest.mark.parametrize('reader', [0, 1, 2])
    # catch Windows environment since we cannot use _read() with custom fast_reader
    @pytest.mark.parametrize("parallel", [False,
        pytest.param(True, marks=pytest.mark.xfail(os.name == 'nt', reason="Multiprocessing is currently unsupported on Windows"))])
    def test_data_out_of_range(parallel, reader):
        """
        Numbers with exponents beyond float64 range (|~4.94e-324 to 1.7977e+308|)
        shall be returned as 0 and +-inf respectively by the C parser, just like
        the Python parser.
        Test fast converter only to nominal accuracy.
        """
        # Python reader and strtod() are expected to return precise results
        rtol = 1.e-30
        if reader > 1:
            rtol = 1.e-15
        # passing fast_reader dict with parametrize does not work!
        if reader > 0:
            fast_reader = {'parallel': parallel, 'use_fast_converter': reader > 1}
        else:
            fast_reader = False
        if parallel:
            if reader < 1:
                pytest.skip("Multiprocessing only available in fast reader")
            elif TRAVIS:
                pytest.xfail("Multiprocessing can sometimes fail on Travis CI")
    
        fields = ['10.1E+199', '3.14e+313', '2048e+306', '0.6E-325', '-2.e345']
        values = np.array([1.01e200, np.inf, np.inf, 0.0, -np.inf])
        t = ascii.read(StringIO(' '.join(fields)), format='no_header', guess=False,
                       fast_reader=fast_reader)
        read_values = np.array([col[0] for col in t.itercols()])
        assert_almost_equal(read_values, values, rtol=rtol, atol=1.e-324)
    
        # test some additional corner cases
        fields = ['.0101E202', '0.000000314E+314', '1777E+305', '-1799E+305', '0.2e-323',
                   '2500e-327', ' 0.0000000000000000000001024E+330']
        values = np.array([1.01e200, 3.14e307, 1.777e308, -np.inf, 0.0, 4.94e-324, 1.024e308])
        t = ascii.read(StringIO(' '.join(fields)), format='no_header', guess=False,
                       fast_reader=fast_reader)
        read_values = np.array([col[0] for col in t.itercols()])
>       assert_almost_equal(read_values, values, rtol=rtol, atol=1.e-324)

astropy/io/ascii/tests/test_c_reader.py:997: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
astropy/io/ascii/tests/common.py:42: in assert_almost_equal
    assert np.allclose(a, b, **kwargs)
/opt/python/cp36-cp36m/lib/python3.6/site-packages/numpy/core/numeric.py:2256: in allclose
    res = all(isclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = array(['1.01e+200', '3.14e+307', '1.777e+308', '-inf', '0.2e-323',
       '5e-324', '1.024e+308'], dtype='<U32')
b = array([1.010e+200, 3.140e+307, 1.777e+308,       -inf, 0.000e+000,
       4.941e-324, 1.024e+308])
rtol = 1e-30, atol = 0.0, equal_nan = False

    def isclose(a, b, rtol=1.e-5, atol=1.e-8, equal_nan=False):
        """
        Returns a boolean array where two arrays are element-wise equal within a
        tolerance.
    
        The tolerance values are positive, typically very small numbers.  The
        relative difference (`rtol` * abs(`b`)) and the absolute difference
        `atol` are added together to compare against the absolute difference
        between `a` and `b`.
    
        Parameters
        ----------
        a, b : array_like
            Input arrays to compare.
        rtol : float
            The relative tolerance parameter (see Notes).
        atol : float
            The absolute tolerance parameter (see Notes).
        equal_nan : bool
            Whether to compare NaN's as equal.  If True, NaN's in `a` will be
            considered equal to NaN's in `b` in the output array.
    
        Returns
        -------
        y : array_like
            Returns a boolean array of where `a` and `b` are equal within the
            given tolerance. If both `a` and `b` are scalars, returns a single
            boolean value.
    
        See Also
        --------
        allclose
    
        Notes
        -----
        .. versionadded:: 1.7.0
    
        For finite values, isclose uses the following equation to test whether
        two floating point values are equivalent.
    
         absolute(`a` - `b`) <= (`atol` + `rtol` * absolute(`b`))
    
        The above equation is not symmetric in `a` and `b`, so that
        `isclose(a, b)` might be different from `isclose(b, a)` in
        some rare cases.
    
        Examples
        --------
        >>> np.isclose([1e10,1e-7], [1.00001e10,1e-8])
        array([True, False])
        >>> np.isclose([1e10,1e-8], [1.00001e10,1e-9])
        array([True, True])
        >>> np.isclose([1e10,1e-8], [1.0001e10,1e-9])
        array([False, True])
        >>> np.isclose([1.0, np.nan], [1.0, np.nan])
        array([True, False])
        >>> np.isclose([1.0, np.nan], [1.0, np.nan], equal_nan=True)
        array([True, True])
        """
        def within_tol(x, y, atol, rtol):
            with errstate(invalid='ignore'):
                return less_equal(abs(x-y), atol + rtol * abs(y))
    
        x = asanyarray(a)
        y = asanyarray(b)
    
        # Make sure y is an inexact type to avoid bad behavior on abs(MIN_INT).
        # This will cause casting of x later. Also, make sure to allow subclasses
        # (e.g., for numpy.ma).
        dt = multiarray.result_type(y, 1.)
        y = array(y, dtype=dt, copy=False, subok=True)
    
>       xfin = isfinite(x)
E       TypeError: ufunc 'isfinite' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```
</details>

I investigated this and the issue is that the value ``0.2e-323`` is not getting converted to a float on 32-bit (the conversion fails and it remains a string, causing the type error above). It turns out that in that 32-bit Docker image, the following code:
 
```c
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>

int main() {
  char *tmp;
  strtod("0.2e-323", &tmp);
  printf("errno=%d ERANGE=%d EDOM=%d\n", errno, ERANGE, EDOM);
}
```

returns:

```
errno=33 ERANGE=34 EDOM=33
```

whereas on other machines it is:

```
errno=34 ERANGE=34 EDOM=33
```

This means that on other machines, the error is an ``ERANGE`` and the value is just set to zero, whereas on the 32-bit machine, the error is ``EDOM``, which causes a ``ValueError`` to be raised here: https://github.com/astropy/astropy/blob/master/astropy/io/ascii/cparser.pyx#L735

There are several solutions here:

* Skip that part of the test on CircleCI

* Skip that part of the test on 32-bit systems

* Change the value so it doesn't cause an error on the 32-bit machine

For anyone who wants to reproduce this, you can use the docker image ``quay.io/pypa/manylinux1_i686``

@taldcroft - what do you think would be best?

**Comment by olebole** [comment]
Since I will run the tests during build time of i386 Debian packages, it would be nice to not just disable it on CircleCI.

**Comment by astrofrog** [comment]
It turns out that just ``0.1e-323`` and ``0.2e-323`` cause the issue, not ``0.4e-323``, so I've just changed the value in #7126

**Comment by astrofrog** [comment]
Fixed in #7126
